File: src/main/java/ru/ispras/microtesk/test/preparator/TestData.java
/*
 * Copyright (c) 2014 ISPRAS (www.ispras.ru)
 * 
 * Institute for System Programming of Russian Academy of Sciences
 * 
 * 25 Alexander Solzhenitsyn st. Moscow 109004 Russia
 * 
 * All rights reserved.
 * 
 * TestData.java, Sep 30, 2014 5:28:51 PM Andrei Tatarnikov
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package ru.ispras.microtesk.test.preparator;

import ru.ispras.microtesk.model.api.data.Data;
import ru.ispras.microtesk.test.template.Primitive;

public final class TestData
{
    public static TestData newTestData(
        String argumentName, Data value)
    {
        if (null == value)
            throw new NullPointerException();

        if (null == argumentName)
            throw new NullPointerException();

        return new TestData(
            value, TargetKind.IMM, null, argumentName);
    }

    public static TestData newTestData(
        String argumentName, Primitive mode, Data value)
    {
        if (null == value)
            throw new NullPointerException();

        if (null == argumentName)
            throw new NullPointerException();

        if (Primitive.Kind.MODE != mode.getKind())
            throw new IllegalArgumentException();

        return new TestData(
            value, TargetKind.MODE, mode, argumentName);
    }

    public static TestData newTestData(
        Location location, Data value)
    {
        if (null == location)
            throw new NullPointerException();

        if (null == value)
            throw new NullPointerException();

        return new TestData(
            value, TargetKind.LOCATION, location, null);
    }

    public static enum TargetKind
    {
        IMM      (null),            // Immediate value
        MODE     (Primitive.class), // Addressing mode
        LOCATION (Location.class);  // Location (MEM, REG, VAR)

        private final Class<?> c;
        private TargetKind(Class<?> c) { this.c = c; }

        private void checkClass(Object target)
        {
            if ((null == c) && (null == target)) return;
            if (c == target.getClass()) return;
            throw new IllegalArgumentException();
        }
    }

    private final Data value;
    private final TargetKind targetKind;
    private final Object target;
    private final String argumentName;

    public TestData(
        Data value, TargetKind kind, Object target, String argumentName)
    {
        if (null == value)
            throw new NullPointerException();

        if (null == kind)
            throw new NullPointerException();

        kind.checkClass(target);

        this.value = value;
        this.targetKind = kind;
        this.target = target;
        this.argumentName = argumentName;
    }

    public Data getValue()
    {
        return value;
    }

    public TargetKind getTargetKind()
    {
        return targetKind;
    }

    public Object getTarget()
    {
        return target;
    }

    public String getArgumentName()
    {
        return argumentName;
    }
}

File: src/main/arch/demo/minimips/templates/testdata.rb
#
# Copyright 2018 ISP RAS (http://www.ispras.ru)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require_relative 'minimips_base'

#
# Description:
#
# This test template demonstrates how to use test data generators that are
# independent of the context and produce multiple sets of test data.
#
class TestDataTemplate < MiniMipsBaseTemplate

  def pre
    super

    #
    # Test case level epilogue
    #
    epilogue {
      nop
    }

    #
    # Start address
    #
    org 0x00020000
  end

  def run
    # Test data for individual registers.
    sequence(:data_combinator => 'product') {
      add reg(8),  (reg(9)  do testdata('range', :min => 1, :max => 3) end),
                   (reg(10) do testdata('range', :min => 1, :max => 3) end)

      sub reg(11), (reg(12) do testdata('range', :min => 1, :max => 3) end),
                   (reg(13) do testdata('range', :min => 1, :max => 3) end)
    }.run

    # Test data for entire instructions.
    sequence(:data_combinator => 'product') {
      add reg(8),  reg(9),  reg(10) do testdata('range', :min => 1, :max => 3) end
      sub reg(11), reg(12), reg(13) do testdata('range', :min => 1, :max => 3) end
    }.run
  end

end
